crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -1105,6 +1111,71 @@ fn find_only_token_in_range(
     token
 }
 
+/// Attach an enclosed end-of-line comment to a [`StmtImportFrom`].
+///
+/// For example, given:
+/// ```python
+/// from foo import (  # comment
+///    bar,
+/// )
+/// ```
+///
+/// The comment will be attached to the `StmtImportFrom` node as a dangling comment, to ensure
+/// that it remains on the same line as the `StmtImportFrom` itself.
+fn handle_enclosed_import_from_comment<'a>(
+    comment: DecoratedComment<'a>,
+    import_from: &'a StmtImportFrom,
+    locator: &Locator,
+) -> CommentPlacement<'a> {
+    let node = comment.enclosing_node();
+
+    // The comment needs to be on the same line, but before the first member, to triggering on:
+    // ```python
+    // from foo import (bar, baz,  # comment
+    //     qux,
+    // )
+    // ```
+    if comment.line_position().is_end_of_line()
+        && !locator.contains_line_break(TextRange::new(import_from.start(), comment.start()))
+        && import_from
+            .names
+            .first()
+            .map_or(false, |name| comment.end() < name.start())
+    {
+        CommentPlacement::dangling(node, comment)
+    } else {
+        CommentPlacement::Default(comment)
+    }
+}
+
+/// Attach a trailing end-of-line comment to a [`StmtImportFrom`].
+///
+/// For example, given:
+/// ```python
+/// from foo import bar  # comment
+/// ```
+///
+/// The comment will be attached to the `StmtImportFrom` node as a dangling comment, to avoid
+/// moving the comment "off" of the `StmtImportFrom` node if it's expanded to multiple lines.
+fn handle_trailing_import_from_comment<'a>(
+    comment: DecoratedComment<'a>,
+    import_from: &'a StmtImportFrom,
+    locator: &Locator,
+) -> CommentPlacement<'a> {
+    let Some(node) = comment.preceding_node() else {
+        return CommentPlacement::Default(comment);
+    };
+
+    // The comment needs to be on the same line.
+    if comment.line_position().is_end_of_line()
+        && !locator.contains_line_break(TextRange::new(import_from.start(), comment.start()))
+    {
+        CommentPlacement::dangling(node, comment)
+    } else {
+        CommentPlacement::Default(comment)
+    }
+}
```

Comment:
I'm a bit surprised that this isn't working by default. The default comment placement associates the comment with the most outer node:

```
# comment
a + b
```

The default placement associates the comment with the `ExpressionStatement` and not the `Expr::Name`, because that would expand the binary expression. The same should be true for trailing comments. 

I pasted the given example and the default placement logic does place the comment correctly:

```
[crates/ruff_python_formatter/src/lib.rs:316] formatted.context().comments().debug(formatted.context().source_code()) = {
    Node {
        kind: StmtImportFrom,
        range: 1..20,
        source: `from foo import bar`,
    }: {
        "leading": [],
        "dangling": [],
        "trailing": [
            SourceComment {
                text: "# comment",
                position: EndOfLine,
                formatted: true,
            },
        ],
    },
}
```
Could you update the example or remove the placement logic if it is unnecessary

Reply:
The default logic does not work here. Given `from foo import bar  # comment`, we treat `# comment` as a trailing comment on `from foo import bar`. However, that doesn't do the right thing.

Given:
```python
from foo import bar, bar, bar, bar, bar, bar, bar, bar, bar, bar, bar, bar, bar, bar, bar, bar  # comment
```

_Without_ this code, we format this as:
```python
from foo import (
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
)  # comment
```

Which is incorrect. We need to keep the `# comment` attached to the `import`.

Instead, we want to do:

```python
from foo import (  # comment
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
    bar,
)
```

So we treat these comments as dangling and handle the placement in the statement formatter.

Reply:
It looks like Black gets this wrong too so I will revert this: https://github.com/astral-sh/ruff/pull/6216#discussion_r1280603568. I thought I tested it yesterday but I must've been mixing up all the different cases.

Reply:
Can we update the method comment? It seems misleading to have an example that doesn't require any special handling.

Reply:
I'm just gonna revert for now, for simplicity, even though I think our behavior is slightly better.